> ### **Assignment 2 - Numpy Array Operations** 
>
> This assignment is part of the course ["Data Analysis with Python: Zero to Pandas"](http://zerotopandas.com). The objective of this assignment is to develop a solid understanding of Numpy array operations. In this assignment you will:
> 
> 1. Pick 5 interesting Numpy array functions by going through the documentation: https://numpy.org/doc/stable/reference/routines.html 
> 2. Run and modify this Jupyter notebook to illustrate their usage (some explanation and 3 examples for each function). Use your imagination to come up with interesting and unique examples.
> 3. Upload this notebook to your Jovian profile using `jovian.commit` and make a submission here: https://jovian.ml/learn/data-analysis-with-python-zero-to-pandas/assignment/assignment-2-numpy-array-operations
> 4. (Optional) Share your notebook online (on Twitter, LinkedIn, Facebook) and on the community forum thread: https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575 . 
> 5. (Optional) Check out the notebooks [shared by other participants](https://jovian.ml/forum/t/assignment-2-numpy-array-operations-share-your-work/10575) and give feedback & appreciation.
>
> The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.
>
> Try to give your notebook a catchy title & subtitle e.g. "All about Numpy array operations", "5 Numpy functions you didn't know you needed", "A beginner's guide to broadcasting in Numpy", "Interesting ways to create Numpy arrays", "Trigonometic functions in Numpy", "How to use Python for Linear Algebra" etc.
>
> **NOTE**: Remove this block of explanation text before submitting or sharing your notebook online - to make it more presentable.


# Exploring the Power of NumPy: A Deep Dive into Essential Functions




NumPy (Numerical Python) is a fundamental library for scientific computing in Python. It provides powerful tools for working with arrays and matrices of numeric data, making it an essential tool for many scientific and data analysis tasks 

- add(x1, x2)
- concatenate([axis, out, dtype, casting])
- sort(a[, axis, kind, order])
- count(a, sub[, start, end])
- sqrt(x)

The recommended way to run this notebook is to click the "Run" button at the top of this page, and select "Run on Binder". This will run the notebook on mybinder.org, a free online service for running Jupyter notebooks.

In [1]:
!pip install jovian --upgrade -q

In [2]:
import jovian

In [3]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Updating notebook "manishmv/numpy-array-operations" on https://jovian.com
[jovian] Committed successfully! https://jovian.com/manishmv/numpy-array-operations


'https://jovian.com/manishmv/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [4]:
import numpy as np

In [5]:
# List of functions explained 
function1 = np.char.add(x1, x2)  
function2 = np.concatenate((a1, a2, ...), axis=0, out=None, dtype=None, casting="same_kind")
function3 = np.sort(a, axis=-1, kind=None, order=None)
function4 = np.char.count(a, sub, start=0, end=None)
function5 = np.emath.sqrt(x)

NameError: name 'x1' is not defined

## Function 1 - np.char.add(x1, x2)

The np.char.add(x1, x2) function in NumPy is used to concatenate two string arrays element-wise. 
It takes two string arrays x1 and x2 as input and returns a new array with concatenated elements

In [9]:
# Example 1 - working 
x1 = ['Hello','Good','Have a great']
        
x2 = [' People',' Morning!',' Day']

np.char.add(x1,x2)

array(['Hello People', 'Good Morning!', 'Have a great Day'], dtype='<U21')

We had two arrays namely x1 and x2. x1 had three words and x2 had three words. the 'np.char.add(x1,x2)' function returns an array with element wise concatenation from the two array x1 and x2.

In [10]:
# Example 2 - working

y1 = ['apple', 'banana', 'cherry']

y2 = [' juice', ' smoothie', ' pie']

np.char.add(y1,y2)

array(['apple juice', 'banana smoothie', 'cherry pie'], dtype='<U15')

Just like the previous example, the corresponding elements from the two arrays y1 and y2 are concatenated in the resulatant array after using the function np.char.add(y1,y2)

In [11]:
# Example 3 - breaking (to illustrate when it breaks)
arr1 = ['Hello','The World']

arr2 = [' People',' is beautiful','Good evening']

np.char.add(arr1,arr2)

ValueError: shape mismatch: objects cannot be broadcast to a single shape

As we can observe from the above example, The add() of numpy strings cannot be applied if there are different number of elements in the two arrays given as arguement, and hence results in an error. In other words the Shape of the two arrays must be same to concatenate their corresponding elements.

The above function can be used to Concatenating two string arrays and Appending a string to each element of an array.

In [12]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "manishmv/numpy-array-operations" on https://jovian.com
[jovian] Committed successfully! https://jovian.com/manishmv/numpy-array-operations


'https://jovian.com/manishmv/numpy-array-operations'

## Function 2 - np.concatenate((a1, a2, ...), axis=0, out=None, dtype=None, casting="same_kind")

The np.concatenate() function in NumPy is used to concatenate two or more arrays along a specified axis. It takes as input a tuple or list of arrays to be concatenated and returns a single array containing the concatenated elements along the specified axis.

In [13]:
# Example 1 - working

x = np.array([[1, 2], [3, 4]])
y = np.array([[5, 6], [7, 8]])

np.concatenate((x, y), axis=0)

array([[1, 2],
       [3, 4],
       [5, 6],
       [7, 8]])

In this example, we use the np.concatenate() function to concatenate two arrays x and y along the rows (axis=0). The resulting array contains all the rows of x and y concatenated together

In [19]:
# Example 2 - working

x = np.array([[1, 2], [3, 4]])
y = np.array([[5, 6], [7, 8]])

np.concatenate((x, y), axis=1)

array([[1, 2, 5, 6],
       [3, 4, 7, 8]])

In this example, we use the np.concatenate() function to concatenate two arrays x and y along the columns (axis=1). The resulting array contains all the columns of x and y concatenated together.

In [20]:
# Example 3 - breaking (to illustrate when it breaks)

x = np.array([[1, 2], [3, 4]])
y = np.array([[5, 6, 7], [8, 9, 10]])

np.concatenate((x, y), axis=0)

ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 2 and the array at index 1 has size 3

In this example, we are trying to concatenate two arrays x and y along the rows (axis=0). However, the number of rows in both arrays do match, but the number of columns do not match. As a result, the np.concatenate() function raises a ValueError saying that all the input array dimensions except for the concatenation axis must match exactly.

So, it's important to ensure that the dimensions of the arrays you are concatenating along the specified axis match exactly, otherwise the np.concatenate() function may not give you the expected results.


In [21]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "manishmv/numpy-array-operations" on https://jovian.com
[jovian] Committed successfully! https://jovian.com/manishmv/numpy-array-operations


'https://jovian.com/manishmv/numpy-array-operations'

## Function 3 - sort(a[, axis, kind, order])

The np.sort() function in NumPy is used to return a sorted copy of an input array along a specified axis. It takes as input an array to be sorted and returns a new array with the same shape as the original array, but with the elements sorted in ascending order by default.

In [22]:
# Example 1 - working


arr = np.array([3, 1, 4, 2, 5])
np.sort(arr)


array([1, 2, 3, 4, 5])

In this example, we use the np.sort() function to sort a 1D array arr in ascending ord

In [23]:
# Example 2 - working

arr = np.array([[3, 1], [4, 2], [5, 6]])
sorted_arr = np.sort(arr, axis=0)[::-1]

print(sorted_arr)

[[5 6]
 [4 2]
 [3 1]]


In this example, we use the np.sort() function to sort a 2D array arr along the columns in descending order. The resulting sorted array sorted_arr contains the same elements as the original array, but sorted in descending order along the columns

In [24]:
# Example 3 - breaking (to illustrate when it breaks)

arr = np.array([[3, 1], [4, 2], [5, 6]])
sorted_arr = np.sort(arr, axis=(0,1))

print(sorted_arr)


TypeError: an integer is required (got type tuple)

In this example, we try to sort a 2D array arr along both rows and columns using the axis parameter set to (0,1). However, this results in an AxisError since the axis parameter does not support multiple axes for sorting. We can only sort along a single axis at a time.

In [25]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "manishmv/numpy-array-operations" on https://jovian.com
[jovian] Committed successfully! https://jovian.com/manishmv/numpy-array-operations


'https://jovian.com/manishmv/numpy-array-operations'

## Function 4 - count(a, sub[, start, end])

The np.char.count() function is used to count the number of non-overlapping occurrences of a substring in a string or an array of strings. It takes as input a string or an array of strings, along with a substring to search for, and returns the count of occurrences of the substring in each element of the input array

In [26]:
# Example 1 - working

str = 'hello world'
count = np.char.count(str, 'l')

print(count)


3


In this example, we use the np.char.count() function to count the number of occurrences of the substring 'l' in a string str. The resulting count is 3 since there are three non-overlapping occurrences of the substring 'l' in the string.

In [27]:
# Example 2 - working

arr = np.array(['hello', 'world', 'hello world'])
count = np.char.count(arr, 'o')

print(count)


[1 1 2]


In this example, we use the np.char.count() function to count the number of occurrences of the substring 'o' in an array of strings arr. The resulting count is an array [1 1 2] that contains the count of occurrences of the substring 'o' in each element of the input array.

In [28]:
# Example 3 - breaking (to illustrate when it breaks)

arr = np.array([1, 2, 3])
count = np.char.count(arr, 'o')

print(count)


TypeError: string operation on non-string array

In this example, we try to use the np.char.count() function on a non-string array arr. This results in an AttributeError since the function only works on string or character arrays.

In [29]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "manishmv/numpy-array-operations" on https://jovian.com
[jovian] Committed successfully! https://jovian.com/manishmv/numpy-array-operations


'https://jovian.com/manishmv/numpy-array-operations'

## Function 5 - np.emath.sqrt(x)

The np.emath.sqrt(x) function is used to calculate the square root of a complex number or an array of complex numbers. It returns an array of complex numbers with the square root of each element in the input array

In [30]:
# Example 1 - working

z = 2 + 3j
sqrt_z = np.emath.sqrt(z)

print(sqrt_z)

(1.6741492280355401+0.8959774761298381j)


In this example, we use the np.emath.sqrt() function to calculate the square root of a single complex number z = 2 + 3j. The resulting square root is a complex number (1.6741492280355401+0.8959774761298381j).

In [31]:
# Example 2 - working

arr = np.array([9,16,25])
sqrt_arr = np.emath.sqrt(arr)

print(sqrt_arr)

[3. 4. 5.]


In this example, we use the np.emath.sqrt() function to calculate the square root of an array of numbers arr

In [37]:
# Example 3 - breaking (to illustrate when it breaks)


x = 'abc'
sqrt_x = np.emath.sqrt(x)

print(sqrt_x)


TypeError: '<' not supported between instances of 'numpy.ndarray' and 'int'

Here we have given a string as value and hence we don't get any square root back as result.

In [38]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "manishmv/numpy-array-operations" on https://jovian.com
[jovian] Committed successfully! https://jovian.com/manishmv/numpy-array-operations


'https://jovian.com/manishmv/numpy-array-operations'

## Conclusion

In this notebook we have learnt about five different functions in NumPy and along with that examples of the working of functions as well as examples when the function doesn't work. This Notebook serves as a great resource for beginners

## Reference Links
references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* ...

In [40]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "manishmv/numpy-array-operations" on https://jovian.com
[jovian] Committed successfully! https://jovian.com/manishmv/numpy-array-operations


'https://jovian.com/manishmv/numpy-array-operations'